x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/movingPercentiles/MovingPercentilesPipelineAggregator.java

```diff
@@ -0,0 +1,276 @@
+                                                   InternalMultiBucketAggregation.InternalBucket bucket,
+                                                   String aggPath) {
+        List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
+        Object propertyValue = bucket.getProperty(agg.getName(), aggPathsList);
+        if (propertyValue == null) {
+            throw new AggregationExecutionException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+                + " must reference a TDigest percentile aggregation");
+        }
+
+        if (propertyValue instanceof InternalTDigestPercentiles) {
+            return ((InternalTDigestPercentiles) propertyValue).getState();
+        }
+
+        String currentAggName;
+        if (aggPathsList.isEmpty()) {
+            currentAggName = agg.getName();
+        } else {
+            currentAggName = aggPathsList.get(0);
+        }
+
+        throw new AggregationExecutionException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+            + " must reference a TDigest percentiles aggregation, got: ["
+            + propertyValue.getClass().getSimpleName() + "] at aggregation [" + currentAggName + "]");
+    }
+
+    private DoubleHistogram resolveHDRBucketValue(MultiBucketsAggregation agg,
+                                                  InternalMultiBucketAggregation.InternalBucket bucket,
+                                                  String aggPath) {
+        List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
+        Object propertyValue = bucket.getProperty(agg.getName(), aggPathsList);
+        if (propertyValue == null) {
+            throw new AggregationExecutionException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+                + " must reference a HDR percentile aggregation");
+        }
+
+        if (propertyValue instanceof InternalHDRPercentiles) {
+            return ((InternalHDRPercentiles) propertyValue).getState();
+        }
+
+        String currentAggName;
+        if (aggPathsList.isEmpty()) {
+            currentAggName = agg.getName();
+        } else {
+            currentAggName = aggPathsList.get(0);
+        }
+
+        throw new AggregationExecutionException(AbstractPipelineAggregationBuilder.BUCKETS_PATH_FIELD.getPreferredName()
+            + " must reference a HDR percentiles aggregation, got: ["
+            + propertyValue.getClass().getSimpleName() + "] at aggregation [" + currentAggName + "]");
+    }
+
+    private int clamp(int index, int length) {
+        if (index < 0) {
+            return 0;
+        }
+        if (index > length) {
+            return length;
+        }
+        return index;
+    }
+
+    /** helper class to collect the percentile's configuration */
+    private static class PercentileConfig {
```

Comment:
Let's add a TODO here to replace this with the `PercentilesConfig` that's used by the percentiles builder.  I don't think we can do that yet since the config isn't available through the Internal objects... but eventually when we refactor pipelines sufficiently they should be able to grab this from original request builder object I think.